source/agora/consensus/validation/Enrollment.d

```diff
@@ -0,0 +1,217 @@
+/*******************************************************************************
+
+    Contains validation rountines for enrollments
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.validation.Enrollment;
+
+import agora.common.Amount;
+import agora.common.Types;
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.UTXOSet;
+
+import std.conv;
+
+version (unittest)
+{
+    import agora.common.Hash;
+    import agora.consensus.data.Transaction;
+}
+
+/*******************************************************************************
+
+    Check the validity of an enrollment.
+
+    A Validator's enrollment is considered valid if:
+        - UTXO is unspent frozen utxo
+        - Signatures are authentic
+        - The frozen amount must be equal to or greater than 40,000 BOA
+
+    Params:
+        enrollment = The enrollment of the target to be verified
+        findUTXO = delegate to find the referenced unspent UTXOs with
+
+    Returns:
+        `null` if the validator's UTXO is valid, otherwise a string
+        explaining the reason it is invalid.
+
+*******************************************************************************/
+
+public string isInvalidReason (const ref Enrollment enrollment,
+    UTXOFinder findUTXO) nothrow @safe
+{
+    UTXOSetValue utxo_set_value;
+    if (!findUTXO(enrollment.utxo_key, size_t.max, utxo_set_value))
+        return "Unspent frozen UTXO not found for the validator.";
+
+    if (utxo_set_value.type != typeof(utxo_set_value.type).Freeze)
+        return "UTXO is not frozen.";
+
+    Point address;
+    try
+    {
+        address = Point(utxo_set_value.output.address);
+    }
+    catch (Exception ex)
+    {
+        return "Error converting address to point";
+    }
+
+    if (!verify(address, enrollment.enroll_sig, enrollment))
+        return "Enrollment signature verification has an error.";
```

Comment:
I know this PR is just moving things around.

Anyway we could reword this a bit, because it sounds like the verification process itself has an error rather than the signature being invalid.

```suggestion
        return "Enrollment signature is invalid";
```

Reply:
There's a couple things I actually want to do after this split. Fixing the wording is one of them, and tiny things like removing full stop. The other is to add some `scope`, etc..
So I think we can leave this for the next PR where we revisit all messages to be consistent, what do you think ?